### ISSUE TITLE
UMAP segfault error

### ISSUE THREAD

**Issue by asmariyaz23** [lmcinnes/umap#265]
I get the following error when running UMAP through Seurat (which uses UMAP)

```
UMAP(a=None, angular_rp_forest=False, b=None, init='spectral',
     learning_rate=1.0, local_connectivity=1, metric='correlation',
     metric_kwds=None, min_dist=0.3, n_components=2, n_epochs=None,
     n_neighbors=30, negative_sample_rate=5, random_state=None,
     repulsion_strength=1.0, set_op_mix_ratio=1.0, spread=1.0,
     target_metric='categorical', target_metric_kwds=None,
     target_n_neighbors=-1, target_weight=0.5, transform_queue_size=4.0,
     transform_seed=42, verbose=True)
Construct fuzzy simplicial set
Fri Jul 12 09:51:05 2019 Finding Nearest Neighbors
Fri Jul 12 09:51:05 2019 Building RP forest with 10 trees
/Users/asmaimran/Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py:469: NumbaWarning: 
Compilation is falling back to object mode WITH looplifting enabled because Function "make_angular_tree" failed type inference due to: Cannot unify RandomProjectionTreeNode(array(int64, 1d, C), bool, none, none, none, none) and RandomProjectionTreeNode(none, bool, array(float32, 1d, C), none, RandomProjectionTreeNode(array(int64, 1d, C), bool, none, none, none, none), RandomProjectionTreeNode(array(int64, 1d, C), bool, none, none, none, none)) for '$14.16', defined at /Users/asmaimran/Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py (476)

File "Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py", line 476:
def make_angular_tree(data, indices, rng_state, leaf_size=30):
    <source elided>

        left_node = make_angular_tree(data, left_indices, rng_state, leaf_size)
        ^

[1] During: resolving callee type: recursive(type(CPUDispatcher(<function make_angular_tree at 0x1b264f440>)))
[2] During: typing of call at /Users/asmaimran/Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py (476)


File "Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py", line 476:
def make_angular_tree(data, indices, rng_state, leaf_size=30):
    <source elided>

        left_node = make_angular_tree(data, left_indices, rng_state, leaf_size)
        ^

  @numba.jit()
/Users/asmaimran/Library/Python/3.7/lib/python/site-packages/numba/compiler.py:725: NumbaWarning: Function "make_angular_tree" was compiled in object mode without forceobj=True.

File "Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py", line 470:
@numba.jit()
def make_angular_tree(data, indices, rng_state, leaf_size=30):
^

  self.func_ir.loc))
/Users/asmaimran/Library/Python/3.7/lib/python/site-packages/numba/compiler.py:734: NumbaDeprecationWarning: 
Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

For more information visit http://numba.pydata.org/numba-doc/latest/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

File "Library/Python/3.7/lib/python/site-packages/umap/rp_tree.py", line 470:
@numba.jit()
def make_angular_tree(data, indices, rng_state, leaf_size=30):
^

  warnings.warn(errors.NumbaDeprecationWarning(msg, self.func_ir.loc))
Fri Jul 12 09:51:05 2019 NN descent for 13 iterations
/Users/asmaimran/Library/Python/3.7/lib/python/site-packages/umap/nndescent.py:92: NumbaPerformanceWarning: 
The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.

To find out why, try turning on parallel diagnostics, see http://numba.pydata.org/numba-doc/latest/user/parallel.html#diagnostics for help.

File "Library/Python/3.7/lib/python/site-packages/umap/utils.py", line 409:
@numba.njit(parallel=True)
def build_candidates(current_graph, n_vertices, n_neighbors, max_candidates, rng_state):
^

  current_graph, n_vertices, n_neighbors, max_candidates, rng_state
/Users/asmaimran/Library/Python/3.7/lib/python/site-packages/numba/compiler.py:588: NumbaPerformanceWarning: 
The keyword argument 'parallel=True' was specified but no transformation for parallel execution was possible.

To find out why, try turning on parallel diagnostics, see http://numba.pydata.org/numba-doc/latest/user/parallel.html#diagnostics for help.

File "Library/Python/3.7/lib/python/site-packages/umap/nndescent.py", line 47:
    @numba.njit(parallel=True)
    def nn_descent(
    ^

  self.func_ir.loc))
	 0  /  13
	 1  /  13
	 2  /  13
Fri Jul 12 09:51:10 2019 Finished Nearest Neighbor Search

 *** caught segfault ***
address 0xfffffffffffffffa, cause 'memory not mapped'

Traceback:
 1: py_call_impl(callable, dots$args, dots$keywords)
 2: umap$fit_transform(as.matrix(x = object))
 3: RunUMAP.default(object = data.use, assay = assay, n.neighbors = n.neighbors,     n.components = n.components, metric = metric, n.epochs = n.epochs,     learning.rate = learning.rate, min.dist = min.dist, spread = spread,     set.op.mix.ratio = set.op.mix.ratio, local.connectivity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, seed.use = seed.use, metric.kwds = metric.kwds,     angular.rp.forest = angular.rp.forest, reduction.key = reduction.key,     verbose = verbose)
 4: RunUMAP(object = data.use, assay = assay, n.neighbors = n.neighbors,     n.components = n.components, metric = metric, n.epochs = n.epochs,     learning.rate = learning.rate, min.dist = min.dist, spread = spread,     set.op.mix.ratio = set.op.mix.ratio, local.connectivity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, seed.use = seed.use, metric.kwds = metric.kwds,     angular.rp.forest = angular.rp.forest, reduction.key = reduction.key,     verbose = verbose)
 5: RunUMAP.Seurat(object = visp, dims = 1:26, min_dist = 0.75)
 6: RunUMAP(object = visp, dims = 1:26, min_dist = 0.75)
```

I am running python3, installed UMAP using pip and running on MacOSX. Any ideas about what maybe happening?

**Comment by lmcinnes** [comment]
I've never seen anything quite like this -- it seems like the parallelization in numba is failing, and that may be causing some deeper issues. Do you know what version of numba is installed?

**Comment by asmariyaz23** [comment]
numba.__version__
'0.44.1'

**Comment by lmcinnes** [comment]
Thanks -- that should be recent enough, and I believe numba has tests that umap works in their system these days, so it's not an issue on their end. In fact it is likely that those a re just warning that come with newer versions of numba. The real problem may be elsewhere. I admit that I am actually completely at a loss as to what is happening here. Any chance it could be related to the R/Python interaction and memory transfer? Is this a particularly large dataset? Is it a large dataset relative to your available memory on the machine?

**Comment by YubinXie** [comment]
I also got this warning with numba 0.44. 
Data: 20k x 21. 
37GB/128GB memory used in the machine.

**Comment by jamesb93** [comment]
I am getting this error too.

The numpy array is 1350 x 21619 in dimensions with float64's. Not huge but not small.

16GB Macbook Pro and the dataset is loaded from JSON into memory. On disk it is about 950mb.

**Comment by mahermassoud** [comment]
I was getting a similar error on Mac OS 10.14.3: `interrupted by signal 11: SIGSEGV`. Was able to fix with 

```
pip install umap-learn==0.3.9
pip install numba==0.45.1
```

I still get deprecation warnings but at least the code runs

**Comment by diazdc** [comment]
I'm also running into a segfault error on CentOS7 with Seurat, although my traceback looks different. What's interesting is that error is only triggered when running UMAP a second time. It has no problems if run once. My system has plenty of RAM (256G). Any ideas as to what could be causing this?

```
 ***  caught segfault ***
 address 0xfffffffffffffff7, cause 'memory not mapped'

Traceback:
1: RcppParallel::setThreadOptions(numThreads = n_threads)
2: uwot(X = X, n_neighbors = n_neighbors, n_components = n_components,     metric = metric,     n_epochs = n_epochs, alpha = learning_rate, scale = scale, init = init, init_sdev = init_sdev,     spread = spread,     min_dist = min_dist, set_op_mix_ratio = set_op_mix_ratio,     local_connectivity = local_connectivity, bandwidth = bandwidth, gamma = repulsion_strength, negative_sample_rate = negative_sample_rate,     a = a, b = b, nn_method = nn_method, n_trees = n_trees, search_k = search_k,     method = "umap", approx_pow = approx_pow, n_threads = n_threads,     n_sgd_threads = n_sgd_threads, grain_size = grain_size, y = y,     target_n_neighbors = target_n_neighbors, target_weight = target_weight,     target_metric = target_metric, pca = pca, pca_center = pca_center,     pcg_rand = pcg_rand, fast_sgd = fast_sgd, ret_model = ret_model,     ret_nn = ret_nn, tmpdir = tempdir(), verbose = verbose)
3: umap(X = object, n_threads = nbrOfWorkers(), n_neighbors = as.integer(x = n.neighbors),     n_components = as.integer(x = n.components), metric = metric,     n_epochs = n.epochs, learning_rate = learning.rate, min_dist = min.dist,     spread = spread, set_op_mix_ratio = set.op.mix.ratio, local_connectivity = local.connectivity,     repulsion_strength = repulsion.strength, negative_sample_rate = negative.sample.rate,     a = a, b = b, verbose = verbose)
4: RunUMAP.default(object = data.use, assay = assay, umap.method = umap.method,     n.neighbors = n.neighbors, n.components = n.components, metric = metric,     n.epochs = n.epochs, learning.rate = learning.rate, min.dist = min.dist,     spread = spread, set.op.mix.ratio = set.op.mix.ratio, local.connectivity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, seed.use = seed.use, metric.kwds = metric.kwds,     angular.rp.forest = angular.rp.forest, reduction.key = reduction.key,     verbose = verbose)
5: RunUMAP(object = data.use, assay = assay, umap.method = umap.method,     n.neighbors = n.neighbors, n.components = n.components, metric = metric,     n.epochs = n.epochs, learning.rate = learning.rate, min.dist = min.dist,     spread = spread, set.op.mix.ratio = set.op.mix.ratio, local.connectivity = local.connectivity,     repulsion.strength = repulsion.strength, negative.sample.rate = negative.sample.rate,     a = a, b = b, seed.use = seed.use, metric.kwds = metric.kwds,     angular.rp.forest = angular.rp.forest, reduction.key = reduction.key,     verbose = verbose)
6: RunUMAP.Seurat(seurat_obj, dims = dims, min.dist = 0.3, spread = 1)
7: RunUMAP(seurat_obj, dims = dims, min.dist = 0.3, spread = 1)`

**Comment by sleighsoft** [comment]
@diazdc Are you using 0.4dev? Could it be a memory mapped numpy file that has trouble?

**Comment by jlmelville** [comment]
@diazdc, the traceback indicates that you aren’t using the Python UMAP, you’re using the [uwot](https://github.com/jlmelville/uwot) R package. Please open an issue there including this traceback but also as much information as you can about what commands you are using, package versions etc. Also, it will be a few days before I am able to look at this.

If you really wanted to use the Python UMAP then the question is why that didn’t happen, but probably the Seurat community is the best place to get answers.

**Comment by lmcinnes** [comment]
Thanks @jlmelville; good catch.

**Comment by diazdc** [comment]
Thanks for the speedy response everyone. Indeed, uwot was causing the problem, and was remedied by calling "umap-learn" as the preffered method. Here's the command in context if anyone else stumbles into this problem: `RunUMAP(seurat_obj, umap.method = "umap-learn", dims = 1:20)`

**Comment by sleighsoft** [comment]
Seems like the inital question is solved with this.
